feat(@formatjs/react-intl)!: Merge chunks in FormatXMLElementFn

czabaj opened the pull request:
relates #1623

This eases definition of `FormatXMLElementFn` for handling RichText in react-intl. See #1623 for reference.

Commit:
feat(@formatjs/react-intl): Merge chunks in FormatXMLElementFnrelates #1623

czabaj commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -20,6 +21,13 @@ import IntlMessageFormat, {
+  | React.ReactNode
+  | FormatXMLElementFn<React.ReactNode, React.ReactNode>
+>;
+
```
I created a custom type for `FormatMessageValues`, I tend to move it into `./types.ts` and reference it in all places, where `values` are used. Is there any notable differences in `values` between Component API and imperative API or could I use the exact same type everywhere?

longlho replied:
yeah this will mess up functional component

czabaj replied:
Custom type for `FormatMessageValues` was discarded..

czabaj commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -72,6 +80,24 @@ function deepMergeFormatsAndSetTimeZone(
+    return acc;
+  }, {});
+}
+
```
Iterate over `values` and patch all `functions` found. Is not very safe, but AFAIK there is no other way to distinguish `FormatXMLElementFn` and currently, there cannot be passed other functions than `FormatXMLElementFn` into `values`.

czabaj commented on packages/react-intl/src/utils.ts:
```diff
@@ -134,3 +134,14 @@ export function getNamedFormat<T extends keyof CustomFormats>(
+  return function () {
+    return formatXMLElementFn(
+      // eslint-disable-next-line prefer-rest-params
+      arguments.length === 1 ? arguments[0] : React.Children.toArray(arguments)
```
Should I use rest instead of `arguments`?

Commit:
fix(@formatjs/react-intl): Remove custom type for `FormatMessageValues`

relates #1623

pyrocat101 commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -141,6 +165,7 @@ export function formatMessage(
   if (!values && message && typeof message === 'string') {
     return message.replace(/'\{(.*?)\}'/gi, `{$1}`);
   }
+  const patchedValues = values && patchFormatXMLElementFn(values);
```
The name `patched` here does not really capture what this really is. Can you add some comments to explain what this is?

czabaj replied:
Well, naming things is hard discipline 😉  I've changed the name to `unapplyFormatXMLElementFnInValues` and created helper function `isFormatXMLElementFn` which shall improve readability.

pyrocat101 commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -72,6 +73,29 @@ function deepMergeFormatsAndSetTimeZone(
+    | FormatXMLElementFn<React.ReactNode, React.ReactNode>
+  >
+): typeof values {
+  return Object.keys(values).reduce((acc: typeof values, k) => {
```
Not that I am worried, but another breaking change to note here is that non-enumerable properties and non-own properties won't be interpolated by the formatter anymore.

Let's make a note in the changelog.

czabaj replied:
It could be refactored to use `for ... in` loop to preserve all enumerable properties including non-own, but I don't see any real use-case for such support. Notice in changelog and docs shall be sufficient. But if you like, I can use `for ... in`.

pyrocat101 replied:
Making a note in the changelog sounds good to me.

pyrocat101 commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -72,6 +73,29 @@ function deepMergeFormatsAndSetTimeZone(
   };
 }
 
+export function patchFormatXMLElementFn(
```
This is not really "patching", it neither mutates the input in-place, nor is it really what's happening under the hood. Can we figure out a better name?

pyrocat101 reviewed (commented):
Thanks for doing this! Can you explain how this change will solve #1467?

czabaj commented:
> Can you explain how this change will solve #1467?
-- @pyrocat101

The error `Each child in a list should have a unique "key"` originates from `FormatXMLElementFn` when there are nested rich-texts used.

There is also `wrapRichTextChunksInFragment` config option, but it does not solve this issue.

There are two possible sources of `Each child...` error:

1. Error `Each child...` originates when user renders result of `formatMessage` call. User declares message with rich-text, the `FormatXMLElementFn` outputs some React element, thus the output of `formatMessage` cannot be concatenated as string and will be passed as an array of chunks to `formatMessage` caller. User can pass `wrapRichTextChunksInFragment` configuration option to wrap such array into `React.Fragment` and gather single React node instead of array of mixed `Array<string, React.Element>` and this will also prevent `Each child...` error when user renders output of `formatMessage`.
1. Error `Each child...` originates in users `FormatXMLElementFn` implementation. User declares message with rich-text and uses rest operator to gather all arguments of `FormatXMLElementFn`. Rest operator collects arguments into an array and user then renders the array in her `FormatXMLElementFn` implementation. If she declared nested rich-text message, she will obtain mixed arguments of `Array<string, React.Element>` in her `FormatXMLElementFn` and she must assign unique keys to the React elements do defeat `Each child` error in her `FormatXMLElementFn` implementation.

This PR aims to solve second issue and also simplifies work with `FormatXMLElementFn`. In essence, `FormatXMLElementFn` in `react-intl` library would not be variadic, users will receive single argument in `FormatXMLElementFn` at all times always. No need for rest operator and no chance to obtain `Each child...` error, for this will be handled for the users in `react-intl` library.

czabaj committed:
refactor: improve naming

pyrocat101 commented on packages/react-intl/src/utils.ts:
```diff
@@ -134,3 +134,14 @@ export function getNamedFormat<T extends keyof CustomFormats>(
   onError(new UnsupportedFormatterError(`No ${type} format named: ${name}`));
 }
+
+export function unapplyFormatXMLElementFn(
```
Can you add a comment here to explain what this function does?

czabaj replied:
Comment added.

pyrocat101 reviewed (approved):
Thanks again for doing this! This is a breaking change and we might have to release major version for it.

czabaj committed:
chore: Comment `unapplyFormatXMLElementFn`

czabaj commented:
> Thanks again for doing this!
-- @pyrocat101 

You are welcome!

longlho commented:
@czabaj If you also change `intl-messageformat` variadic callback formatting fn do you not have to deal w/ the unpatching stuff?

czabaj commented:
> @czabaj If you also change intl-messageformat variadic callback formatting fn do you not have to deal w/ the unpatching stuff?
-- @longlho 

The problem is that `intl-messageformat` is React agnostic, so it cannot perform a merge of arguments which are React nodes.

longlho commented:
That should be fine. This is a breaking change regardless I believe?

longlho commented:
To clarify I think I'm looking for the `(...chunks)` to be `(chunks)` in intl-messageformat as well

pyrocat101 commented:
@czabaj can you update `intl-messageformat` as well?

czabaj commented:
@longlho @pyrocat101 sorry for the late response.

Package `intl-messageformat` calls `FormatXMLElementFn` for example [here](https://github.com/formatjs/formatjs/blob/master/packages/intl-messageformat/src/formatters.ts#L221) and it spreads `parts` array to the `FormatXMLElementFn`. I think we can omit spread and pass an array of parts to the `formatFn` directly which simplifies the API (no need for rest operator on the consumer side) and should be more performant than variadic function with spread and rest operators. There remains problem with `Each child in a list should have a unique "key"` but we can fix that in `react-intl` package.

The last piece of the puzzle is -- If I change the way `intl-messageformat` handles `FormatXMLElementFn`, how it impacts other packages? Do you know about some dependencies of `intl-messageformat` which relies on variadic `FormatXMLElementFn`?

longlho commented:
it'll be a breaking change and we'll bump major version for it

czabaj commented:
@longlho got it 👌  I will work on in this evening (around 16:00 UTC).

czabaj committed:
refactor: FormatXMLElementFn is no more variadic

czabaj commented:
I changed the signature of `FormatXMLElementFn` to non-variadic in `intl-messageformat`. There remains logic in `react-intl` which assigns unique keys to items of the `parts` array. I successfully ran tests in `react-intl` but I'm unsure whether this breaks something else, please verify.

I also rebased this branch to the tip of the master branch and force pushed. I forgot to ask whether someone tries this locally if someone did, sorry for inconveniences.

czabaj commented on packages/intl-messageformat/src/formatters.ts:
```diff
@@ -86,7 +86,7 @@ function mergeLiteral<T>(
 }
 
-function isFormatXMLElementFn<T>(
+export function isFormatXMLElementFn<T>(
```
I exported this function for I need it in `react-intl`. I previously even declared almost exactly the same function in `react-intl`, but sharing is caring 😄

czabaj commented on packages/react-intl/src/formatters/message.ts:
```diff
@@ -72,6 +74,21 @@ function deepMergeFormatsAndSetTimeZone(
   };
 }
 
+export function assignUniqueKeysToFormatXMLElementFnArgument(
```
I will welcome tips for a better name, any thoughts?

longlho committed:
Merge branch 'master' into 1623

longlho commented:
Looks like the build failed @czabaj can u take a look?

czabaj committed:
test(intl-messageformat): change signature of FormatXMLElementFn

czabaj commented:
@longlho I reflected changed signature of FormatXMLElementFn in tests.

I'm having a problem running tests across whole monorepo, it yields errors

> Error: ENOENT: no such file or directory, stat '.../react-intl/test262/test/intl402/DisplayNames'

and

> /.../react-intl/packages/intl-listformat/tests/runner.ts:48
> @formatjs/intl-listformat: const json: TestResult[] = JSON.parse(result.stdout);
> @formatjs/intl-listformat:                                 ^
> @formatjs/intl-listformat: SyntaxError: Unexpected end of JSON input

but running tests per project work OK

longlho commented:
Thanks a lot for your contributions! I've manually merged this in since I have to write more CHANGELOG!
